x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -232,27 +242,44 @@ protected void closeInternal() {
 
         private final RefCountedCloseable<Engine.IndexCommitRef> session;
         private final RefCountedCloseable<IndexInput> input;
+        private final Releasable lockRelease;
 
-        private Reader(RefCountedCloseable<Engine.IndexCommitRef> session, RefCountedCloseable<IndexInput> input) {
+        private Reader(RefCountedCloseable<Engine.IndexCommitRef> session, RefCountedCloseable<IndexInput> input, Releasable lockRelease) {
             this.session = session;
             this.input = input;
-            session.incRef();
-            input.incRef();
+            this.lockRelease = lockRelease;
+            boolean sessionRefIncremented = false;
+            boolean inputRefIncremented = false;
+            try {
```

Comment:
maybe you wanna use something like this:

```Java
List<Closeable> toRelease = new ArrayList<>();
toRelease.add(lockRelease);
try {
 session.incRef()
 toRelease.add(session);
 input.incRef();
 toRelease.clear(); // all done no need to release anything
} finally {
  IOUtils.closeWhileHandlingException(toRelease);
}

Reply:
With more changes related to discussions with Yannick, this code is no longer needed.